Another fake PR for review (do not close!)

egparedes opened the pull request.

davhofer committed:
Created spack package based on pythonOC package

davhofer committed:
Merge branch 'main' of github.com:davhofer/intern-pyspack

davhofer committed:
Remove CHANGELOG.md

davhofer committed:
Update progress report and readme.

davhofer committed:
Checksum new package versions and add to package.py

davhofer committed:
feat!: add module for (simple) parsing of pyproject.toml files and conversion to spack package.py

davhofer committed:
docs: cleanup readme

davhofer committed:
feat: add conversion for markers from pypi-to-spack-package

davhofer committed:
fix: bugfix for display of platform in when_spec; bugfix in handling of markers/returned spec list; improved error handling and code documentation

davhofer committed:
chore: types & formatting, directory structure

davhofer committed:
chore: updater name ideas

davhofer committed:
refactor: move code from pypi-to-spack-pkg into separate module; make actual installable python pkg (flit backend)

davhofer committed:
chore: rename function names to reflect scope in module (internal/external)

davhofer committed:
feat: multiple .tomls for different versions of a pkg can now be converted to a single package.py

davhofer committed:
docs: update CHANGELOG

davhofer committed:
chore: add linting and formatting to config, apply in code

davhofer committed:
chore: init.py formatting

davhofer committed:
chore: clean up import type annotations

davhofer committed:
docs: add wiki to readme

davhofer committed:
test: add tests for external.py module

davhofer committed:
fix: edge cases in external._best_upperbound and _best_lowerbound. For
now, prereleases and dev, post, and local versions are not supported and
will be excluded.

davhofer committed:
chore: update dependencies; adjust config

davhofer committed:
test: add unittests for parse_pyproject; replace some naming normalization functions with spack builtins

davhofer committed:
chore: remove spack dependencies distributed with it; edit PYTHONPATH instead

davhofer committed:
refactor: replace pyproject_metadata with own utilities for parsing
    (adapted from pyproject_metadata).
    Reason: raised Exceptions in some cases which lead to the whole file
    not being parsed; we want fine grained control over parsing of
    fields and handling of errors.
    Parse errors for dependenceis & files are displayed to the user in
    the output as comments.

davhofer committed:
chore: improve error handling for certain functions; cleanup

davhofer committed:
chore: formatting and style

davhofer committed:
docs: update README

davhofer committed:
chore: correct line lengths

davhofer committed:
docs: update CHANGELOG

davhofer committed:
chore: add docstring

davhofer committed:
feat: add module for downloading sdists directly from PyPI

davhofer committed:
feat!: integrate downloading of pkgs from pypi; convert package directly from pypi with 'SpackPyPkg.convert_pkg'

davhofer committed:
docs: update CHANGELOG

davhofer committed:
fix: small code cleanup

davhofer committed:
fix: uncaught key error in _extract_from_tar when pyproject.toml is missing

davhofer committed:
fix: error in condensed_version_list when valid subset is empty; return empty list

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Update pyproject.toml

Co-authored-by: Enrique González Paredes <enriqueg@cscs.ch>

davhofer committed:
Merge pull request #1 from davhofer/main

Code Review: Improve pyproject.toml config

egparedes commented on src/py2spack/__init__.py:
```diff
@@ -0,0 +1,10 @@
+"""py2spack package.
+
+This package allows converting standard python packages to Spack packages.
+"""
```
The first line should be the summary, not  the name of the package (which is already encoded in the folder or file name). 
```suggestion
"""A package to convert Python distribution packages to Spack packages."""
```

All comments like this one about general conventions or style rules apply to all files in the repo, even if I will only the mention it the first time they appear.

egparedes commented on src/py2spack/__init__.py:
```diff
@@ -0,0 +1,10 @@
+
+__version__ = "0.0.1"
+
+__all__ = ["SpackPyPkg"]
```
Don't forget that our coding conventions (Google coding conventions in this case) build on top of PEP 8, the most basic language style convention which says that the special/magic/dunder module variables should be defined before any regular import:    https://peps.python.org/pep-0008/#module-level-dunder-names
```suggestion
__all__ = ["SpackPyPkg"]
__version__ = "0.0.1"

from .main import SpackPyPkg
```
The rationale for this is to set the module metadata before anything else, to avoid possible errors when importing code relying in this metadata being already defined.

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+        msg: str,
+    ):
+        """Initialize error."""
+        super().__init__(msg)
```
The current `__init__` is completely redundant since it's only calling the `__init__` of the superclass. Additionally, the name of the class is too generic and should be more specific, e.g.   
```suggestion
class PyPIQueryError(Exception):   # Or QueryError, or something along these lines
    """Error during version lookup or querying of the PyPI JSON API."""
    pass
    
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+
+
+def _acceptable_version(version: str) -> Optional[vn.Version]:
+    """Check whether version string can be parsed and has correct format."""
```
Either this function is a predicate checking if the argument is a valid version, in which case it should be named something like `is_acceptable_version()` and return a bool, or is a function parsing version strings which returns `None` if there are errors, and then it should be named like `parse_version()` or `parse_supported_version()`. This case looks like the latter, so I'd suggest to rename it and update the docstring.

In general, functions should be named using verbs, and if they return booleans, they should be named as predicate expressions (`is_larger_than()`, `starts_with()`, `has_children()` , ...)

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+
+import io
+import tarfile
+from typing import Any, Dict, List, Optional
```
From python 3.9 builtin types like `dict` and `list` are generic and therefore they should be used directly, since their aliases in typing are deprecated (https://typing.readthedocs.io/en/latest/guides/modernizing.html#generics-in-the-typing-module)
```suggestion
from typing import Any, Optional
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+            return None
+        return v
+    except vn.InvalidVersion:
+        return None
```
In general the control flow of the code should be as linear as possible to make it easier to understand:  
```suggestion
    result = None
    try:
        v = vn.parse(version)
        # do not support post releases of prereleases etc.
        if not (v.pre and (v.post or v.dev or v.local)):
            result = v
    except vn.InvalidVersion:
        pass

    return result
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+        return None
+
+
+class PyPILookup:
```
This class looks like the specific implementation using PyPI API of a more general concept of some "provider" of python distribution packages. Wouldn't it make more sense to define a `Protocol` for the general provider interface and then a concrete implementation of the interface connecting to PyPI? In this way, different providers (e.g. github?) can be used transparently without changing other components.

Basic sketch of the proposal:
```python
class PyProjectProvider(Protocol):

    @abc.abstractmethod
    def get_versions(self, name: str) -> list[vn.Version]:
        ...

    @abc.abstractmethod
    def get_files(
        self, name: str, versions: int | list[vn.Version] | None  = None
    ) -> List[Dict[Any, Any]] | Error:
         ....

class PyPIAPIProvider(DistributionPackagesProvider):
     url: str =  "https://pypi.org/simple/"
     ....

```
Also, I'm not completely sure if `get_files` is the best interface for this. Wouldn't it be cleaner to have public a method to get the `pyproject` files for some version and another to get the actual distributions packages? Of course both could be implemented in terms of a `_get_files()` helper method

Finally, consider the idea of using a parameter for the URL of the API in the concrete PyPI implementation, so the same class is potentially reusable for different PyPI deployments...

davhofer replied:
@egparedes Regarding the `get_files` interface: 
The method currently only provides metadata of the distribution package like filename, download url, hash, etc.
But it would make sense to directly put methods for loading the pyproject.toml file contents etc. here too. And I agree, this could be done for a specific version of the package. 
I think a public method to get the `pyproject` should be enough though (no public method to get the actual distribution package), since we only need the distribution package to get the pyproject.toml data. So this can just be downloaded and extracted in a helper method.

So in summary, my idea would be to do a protocol with public methods `get_versions(name)` which returns a list of available versions (already filtered for only source distributions?) and `get_pyproject(name, version)` which returns a dict with the contents of the pyproject.toml file.

If later we also need to get other files from the source distribution except the `pyproject.toml`, I could also make a more generic method to extract a file from the source distribution?

egparedes replied:
Yes, I agree. Let's keep it simple for now.

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+
+        Available versions are cached.
+        """
+        r = requests.get(
+            f"https://pypi.org/simple/{name}/",
+            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
+            timeout=10,
+        )
+        if r.status_code != 200:
+            if r.status_code == 404:
+                msg = f"Package {name} not found on PyPI (status code 404)"
+                return APIError(msg)
+
+            msg = (
+                f"Error when querying JSON API (status code {r.status_code})."
+                f" Response: {r.text}"
+            )
+            return APIError(msg)
+
+        data: dict = r.json()
+        versions = data["versions"]
+
+        # parse and sort versions
+        parsed_versions = sorted(
+            {vv for v in versions if (vv := _acceptable_version(v))}
+        )
+
+        self.version_cache[name] = parsed_versions
+
+        return data
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
```
The `fucntools` module comes with memoization functionality (`cache` or `lru_cache`) that could be leveraged here to simplify the implementation. Also I would cleanly decouple the handling of the query to the server and the processing of the returned information. For example:

```suggestion
    def _get_data(self, name: str) -> dict | APIError:
        """Load info for the available distribution files from PyPI."""
        r = requests.get(
            f"https://pypi.org/simple/{name}/",
            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
            timeout=10,
        )
        if r.status_code != 200:
            if r.status_code == 404:
                msg = f"Package {name} not found on PyPI (status code 404)"
                return APIError(msg)

            msg = (
                f"Error when querying JSON API (status code {r.status_code})."
                f" Response: {r.text}"
            )
            return APIError(msg)

        data: dict = r.json()
        return data

    @functools.lru_cache
    def get_versions(self, name: str) -> List[vn.Version]:
        """Get acceptable versions for package (from cache if possible)."""
        data = self._get(name)
        if isinstance(data, APIError):   ## I think this behavior is wrong and it should return an error in this case
            self.version_cache[name] = []
        versions = data["versions"]

        # parse and sort versions
        parsed_versions = sorted(
            {vv for v in versions if (vv := _acceptable_version(v))}
        )

        return parsed_versions
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+
+        return self.version_cache[name]
+
+    def get_files(
```
See related comments above.

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+            f for f in files if _known_archive_format(f["filename"])
+        ]
+
+        if len(files_known_format) == 0:
```
The idiomatic Python way of checking if builtin collections are empty is check if they are false values:
```suggestion
        if not files_known_format:
```
Also: https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+                continue
+
+            try:
+                sha256 = f["hashes"]["sha256"]
```
Can you explain to me when the "sha256" hash could be missing? Why is it correct to continue in that case?

davhofer replied:
@egparedes While the `hashes` field is required, it can be that there is no hash in it, or another hash than `sha256`, according to https://peps.python.org/pep-0691/.
In that case I want to simply skip this specific version and try the next one.

davhofer replied:
spack actually also supports other hashes than sha256, so I could in addition also check for those

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+        ext for ext in KNOWN_ARCHIVE_FORMATS if filename.endswith(ext)
+    ]
+
+    if len(extension_list) == 0:
```
```suggestion
    if not extension_list:
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+        msg = f"Extension not recognized for: {filename}"
+        return APIError(msg)
+
+    if len(extension_list) == 1:
+        return extension_list[0]
```
This case is redundant, since the default case below will return the same value.

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+    filename: str, pkg_name: str, archive_ext: str
+) -> vn.Version | None:
+    """Parse version from filename and check correct formatting."""
+    prefix = pkg_name + "-"
```
```suggestion
    prefix = f"{pkg_name}-"
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+
+
+def _known_archive_format(filename: str):
+    return any([filename.endswith(ext) for ext in KNOWN_ARCHIVE_FORMATS])
```
There is no need to allocate a list for the result of the comprehensions since it's just a temporary value.
```suggestion
    return any(filename.endswith(ext) for ext in KNOWN_ARCHIVE_FORMATS)
```

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+    return any([filename.endswith(ext) for ext in KNOWN_ARCHIVE_FORMATS])
+
+
+def _download_sdist(url: str) -> io.BytesIO | APIError:
```
This function looks like a general utility without specific business logic. I will move it to some utility module (e.g. `utils.py`)

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,266 @@
+        return APIError(msg)
+
+
+def _extract_from_tar(
```
This function looks like a general utility without specific business logic. I will move it to some utility module (e.g. `utils.py`)

egparedes commented on src/py2spack/loading.py:
I think the name of the module is not very precise. What do you think of something along the lines of  `package_providers`, `project_providers`... ?

egparedes reviewed (commented):
Another part of the review: this times I only reviewed the `loading.py` module, although some of the comments are more general and also apply to other modules. Feel free to add your questions and objections here so we can continue the discussion in this fake PR review.

davhofer committed:
test: add tests for loading.py module and update other tests according to refactoring

davhofer committed:
chore: fix formatting according to updated config

egparedes commented on src/py2spack/loading.py:
```diff
@@ -0,0 +1,257 @@
+
+        Available versions are cached.
+        """
+        r = requests.get(
+            f"https://pypi.org/simple/{name}/",
+            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
+            timeout=10,
+        )
+        if r.status_code != 200:
+            if r.status_code == 404:
+                msg = f"Package {name} not found on PyPI (status code 404)"
+                return APIError(msg)
+
+            msg = (
+                f"Error when querying JSON API (status code {r.status_code})."
+                f" Response: {r.text}"
+            )
+            return APIError(msg)
+
+        data: dict = r.json()
+        versions = data["versions"]
+
+        # parse and sort versions
+        parsed_versions = sorted(
+            {vv for v in versions if (vv := _acceptable_version(v))}
+        )
+
+        self.version_cache[name] = parsed_versions
+
+        return data
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
```
A similar caching functionality is already provided by `functools.lru_cache()` or `functools.cache()`. Also, the parsing of the versions doesn't belong to the utility method contacting the server but to the actual method dealing with versions.

```suggestion
    def _get(self, name: str) -> dict | APIError:
        """Load info for the available distribution files from PyPI.

        Available versions are cached.
        """
        r = requests.get(
            f"https://pypi.org/simple/{name}/",
            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
            timeout=10,
        )
        if r.status_code != 200:
            if r.status_code == 404:
                msg = f"Package {name} not found on PyPI (status code 404)"
                return APIError(msg)

            msg = (
                f"Error when querying JSON API (status code {r.status_code})."
                f" Response: {r.text}"
            )
            return APIError(msg)

        data: dict = r.json()
        return data

    @functools.cache
    def get_versions(self, name: str) -> List[vn.Version]:
        """Get acceptable versions for package (from cache if possible)."""
        data = self._get(name) 
        if isinstance(data, APIError):
            return []   ## <----- I don't think it's right to return an empty list instead of an error here...
        versions = data["versions"]

        # parse and sort versions
        parsed_versions = sorted(
            {vv for v in versions if (vv := _acceptable_version(v))}
        )
        return parsed_versions
```

davhofer committed:
refactor: fix ruff errors and refactor complicated functions

davhofer committed:
fix: indentation error

davhofer committed:
chore: update tests and config according to refactoring & formatting changes

davhofer committed:
Code quality and refactoring

- Implemented code quality and clarity changes according to @egparedes's suggestions. Refactored various methods and classes for improved functionality and clarity
- Also fixed type and formatting/style errors in whole project
- Renamed some modules
- Added pre-commit config
- Updated existing tests to reflect changes

Individual commits: 

* refactor!: improve code quality and clarity
- rename some modules for clarity
- move utilities for downloading and extracting archives to new module utils
- refactor and abstract the pyproject.toml provision process. Create protocol PyProjectProvider, instantiation PyPIProvider
- move method _convert_requirement from core to conversion_tools
- fix formatting and style errors
- fix type errors

* chore: rename module loading to package_providers

* chore: add pre-commit config

* test: update tests to reflect code and api changes

* chore: run pre-commit

* chore: cleanup type: ignore comments

* chore: make convert_requirement public

* chore: implement change requests by @egparedes

* chore: some more code cleanup

* refactor!: method `core._find_dependency_satisfiability_conflicts` now returns a list of conflicts which will then be displayed in the package.py

davhofer committed:
feat!: implement automatic conversion of dependencies, add CLI for converting packages

davhofer committed:
test: update tests and fix small issues in code

davhofer committed:
fix: update  to check for default repo using SPACK_ROOT env. var.

davhofer committed:
docs: update readme

davhofer committed:
chore: add 'test' variant to package

davhofer committed:
chore: rename 'test' extra to 'tests'

davhofer committed:
chore: update spack repo format and check

davhofer committed:
fix: _get_spack_repo can now use the 'spack repo list' command to get available spack repos

davhofer committed:
fix: bug in package_providers._parse_version_from_directory_name, normalize package name

davhofer committed:
test: add docker container github actions for running tests automatically on push

davhofer committed:
test: fix error in github action

davhofer committed:
test: fix error in github action

davhofer committed:
test: fix error in github action

davhofer committed:
chore: cleanup files

davhofer committed:
test: run spack installation tests on push & PR using docker image (#6)


* test: setup docker img with spack installation and run installation and import tests in github action

* test: separate gh action for pytest and installation tests

* chore: rename github actions

davhofer committed:
fix: just skip version if version number cannot be parsed from filename

davhofer committed:
test: add example scikit-build-core projects

davhofer committed:
feat: add GitHubProvider class for converting packages from github (#7)

- implement GitHubProvider class as alternative to PyPIProvider
- currently only supports converting the first, main package from github. All dependencies are resolved through pypi.

davhofer committed:
fix: remove print statement

davhofer committed:
Merge branch 'main' into scikit-build-core-integration

davhofer committed:
fix: compute checksums through GitHub provider

davhofer committed:
chore: fix formatting errors

davhofer committed:
docs: update README

davhofer committed:
chore: update comments/cli output

davhofer committed:
docs: edit comment

davhofer committed:
feat!: handle scikit-build-core backend and cmake simple dependencies (#9)

For packages using the scikit-build-core backend (which relies on cmake), extract non-python dependencies and versions from CMakeLists.txt (in a simple/primitive way). Extracted dependencies are added as comments/suggestions to the package.py file.

* feat!: add simple parsing and conversion of some fields CMakeLists.txt for scikit-build-core packages.

Extracted, proposed dependencies are displayed in the package.py as comments.

Renamed PyProjectProvider to PackageProvider.
Changed PackageProvider API to support reading contents of arbitrary files in sdist.
Moved utilities for interacting with Spack repo/exe to spack_utils.py.

* chore: add cmake_parser dependency

* chore: remove comments

* chore: remove unused import

* fix: correct function import

* chore: update pyproject.toml for new dependency

* fix: cleanup cmake_conversion code

* test: fix imports

davhofer committed:
test: fix containerized installation tests (#10)

* test: fix containerized installation tests

* test: figure out what goes wrong

* test: improve 'run_spack_command' to handle missing spack shell executable

davhofer committed:
fix: change 'spack_utils.get_spack_repo', use builtin 'spack repo list' command, display available repos, and ask user for repo by default

davhofer committed:
fix: bug in GitHubProvider if repo url ends with /

davhofer committed:
fix: bug when converting a github pkg, incorrect checking for existing initial package in spack

davhofer committed:
Add sphinx docs (#12)

* Add docs folder and minimal generation workflow using sphinx with myst-parser and autodoc2

* Add docs related dependencies and configs

---------

Co-authored-by: Enrique Gonzalez Paredes <enriqueg@cscs.ch>

davhofer committed:
docs: update readme to mention building of sphinx docs

davhofer committed:
chore: remove old info from README

davhofer committed:
Update README.md

davhofer committed:
feat: improved cmake dependency handling (#13)

Collect external (non-python) dependencies for scikit-build-core packages by recursively enumerating all included CMakeLists.txt files and extracting dependencies from them (only for the most recent package version). The dependencies are displayed in the package.py file as comments, grouped together by name and displayed with the respective source CMakeLists file & line number for reference.

* fix: recursively traverse subdirectories specified in CMakeLists.txt to get non-python dependencies; only parse CMakeLists for one (most recent) package version

* fix: add source information for cmakelists dependencies

* fix: collect dependencies on the same non-python package/library from CMakeLists in output and display together

* test: add tests for conversion of cmake dependencies
